RDoc-1906_RollingIndexDeployment (#1343)
ravendb/docs: Documentation/5.2/Raven.Documentation.Pages/indexes/rolling-index-deployment.markdown

{NOTE/}

---

{PANEL: Why Rolling Index Deployment}

When heavy-duty indexing is performed in parallel by all cluster nodes, the cluster's 
performance and availability may be reduced.  
The extent of the reduction in cluster conduct depends upon its nodes' resources, 
the scope of the required indexing, and the number of nodes concurrently indexing.  

* **On Site**, dedicating much of all nodes' resources to indexing rather than to processing 
  data and tasks may reduce the cluster's performance.  
* **On the Cloud**, parallel indexing may exhaust the credits available to multiple nodes 
  at the same time and degrade the cluster's availability.  

**Rolling index deployment** ensures that indexes will be created and updated while 
the cluster remains fully available and performant.  

{INFO: }
Parallel indexing may be a better option when there is minor or no database activity.  
{INFO/}

{PANEL/}

{PANEL: How Does It Work}

---

### The Rolling Procedure

Nodes are assigned with the indexing of each database in a linear order, one node at a time.  

1. The cluster assigns indexing to one of its nodes.  

> 2. When the assigned node finishes indexing, it send the cluster leader a confirmation 
>    that indexing is done.  

Reviewer: ```suggestion
2. When the assigned node finishes indexing, it send a cluster-wide confirmation command
   that indexing is done.  
```

Author: done

> 3. The cluster leader assigns indexing to the next node.  

Reviewer: ```suggestion
3. The cluster assigns indexing to the next node.  
```

Author: done

   {WARNING: }

>    If the delivery of an **indexing completion confirmation** fails, e.g. because the 
>    sending (indexing) or receiving (cluster leader) node has been forcefully disconnected 
>    from the cluster, **indexing will pend** for all nodes until the cluster recovers or 
>    indexing is initiated manually.  

Reviewer: ```suggestion
   If the delivery of an **indexing completion confirmation** fails,  **indexing will pend** for all nodes until the confirmation succeed or 
   indexing is initiated manually.  
```

Author: If the delivery of an **indexing completion confirmation** fails, **indexing will pend** for all nodes until the confirmation succeeds or indexing is initiated manually.  
Confirmation delivery may fail, for example, due to forceful disconnection of the indexing node or cluster leader node during indexation.

Reviewer: Actually to say that "indexing will pend for all nodes" is not accurate.
What will happened is that the running index will complete its job, but will be unable to notify about that. 
So no other index can start running

Author: done:

If the delivery of an **indexing completion confirmation** fails when the current node finishes indexing, no other node will be able to start indexing until the confirmation succeeds or indexing is initiated manually.

   {WARNING/}

---

### Deployment Scope and Course

* **Rolling Scope**  
  The cluster maintains an **independent indexing deployment course** for each databases.  
* **Rolling Order**  

>   Rolling order is determined by the cluster (node `A` **not** necessarily always being the first).  

Reviewer: The initial order is the reverse order of the Member nodes of the database group concatenate with Rehabs and lastly with the Promotables

Author: * **Rolling Order**  
  The rolling order is determined by the cluster.  
  Indexing is deployed to nodes in the reverse order of their membership in the database group, giving a lower priority to nodes in [Rehab or Promotable](../server/clustering/distribution/distributed-database#database-topology) state.

The implication of this design is that different nodes **can** index different databases concurrently.  

* E.g. -  
   * The indexing deployment course the cluster chose for the **"Integration"** database, is:  
     indexing by node `B`, then by node `C`, and finally by node `A`.  
   * The indexing deployment course the cluster chose for the **"Production"** database, is:  
     indexing by node `C`, then by node `A`, and finally by node `B`.  
   * Assuming that the cluster has started indexing the two databases at the same time,  

>      node `B` will be indexing "Integration" while node `C` indexes "Production".  

Reviewer: If find it hard to follow... In general, each _index_ can have it's own order of deployment

Author: removed. the section is now "Deployment Concurrency and Order", and this part is:

* **Deployment Concurrency**  
   * A node is assigned to index a database, only if no other node currently indexes this database.  
   * Multiple nodes **can** be assigned to concurrently index **different databases**.  
     E.g., node `A` can index the "Integration" database, while node `B` indexes the "Production" database.
